src/redis_pubsub.rs

```diff
@@ -0,0 +1,165 @@
+use redis::{Client, Commands, RedisError};
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::thread;
+use std::time::Duration;
+
+static ERROR_CHANNEL: &'static str = "RUSTG_REDIS_ERROR_CHANNEL";
+
+thread_local! {
+    static REQUEST_SENDER: RefCell<Option<flume::Sender<PubSubRequest>>> = RefCell::new(None);
+    static RESPONSE_RECEIVER: RefCell<Option<flume::Receiver<PubSubResponse>>> = RefCell::new(None);
+}
+
+enum PubSubRequest {
+    Subscribe(String),
+    Publish(String, String),
+}
+
+// response might not be a good name, since those are not sent in response to requests
+enum PubSubResponse {
+    Disconnected(String),
+    Message(String, String),
+}
+
+fn handle_redis_inner(
+    client: Client,
+    control: &flume::Receiver<PubSubRequest>,
+    out: &flume::Sender<PubSubResponse>,
+) -> Result<(), RedisError> {
+    let mut conn = client.get_connection()?;
+    let mut pub_conn = client.get_connection()?;
+    let mut pubsub = conn.as_pubsub();
+    let _ = pubsub.set_read_timeout(Some(Duration::from_secs(1)));
+
+    loop {
+        loop {
+            match control.try_recv() {
+                Ok(req) => match req {
+                    PubSubRequest::Subscribe(channel) => {
+                        pubsub.subscribe(&[channel.as_str()])?;
+                    }
+                    PubSubRequest::Publish(channel, message) => {
+                        pub_conn.publish(&channel, &message)?;
+                    }
+                },
+                Err(flume::TryRecvError::Empty) => break,
+                Err(flume::TryRecvError::Disconnected) => return Ok(()),
+            }
+        }
+
+        if let Some(msg) = match pubsub.get_message() {
+            Ok(msg) => Some(msg),
+            Err(e) => {
+                if e.is_timeout() {
+                    None
+                } else {
+                    return Err(e);
+                }
+            }
+        } {
+            let chan = msg.get_channel_name().to_owned();
+            let data: String = msg.get_payload().unwrap_or_default();
+            if let Err(flume::TrySendError::Disconnected(_)) =
+                out.try_send(PubSubResponse::Message(chan, data))
+            {
+                return Ok(()); // If no one wants to receive any more messages from us, we exit this thread
+            }
+        }
+    }
+}
+
+fn handle_redis(
+    client: Client,
+    control: flume::Receiver<PubSubRequest>,
+    out: flume::Sender<PubSubResponse>,
+) {
+    if let Err(e) = handle_redis_inner(client, &control, &out) {
+        let _ = out.try_send(PubSubResponse::Disconnected(e.to_string()));
```

Comment:
Why are we silencing potential errors here? This is spooky.

Reply:
This could fail if the receiving end was dropped, in which case there's no need (or way) to communicate back the error, or if the channel was full, in which case the DM side was probably FUBAR. I suppose it doesn't hurt to make it block until there's space or if the receiver is dropped.

Reply:
If this blocks, how is the queue ever going to clear if DM cant execute anything to clear the queue due to being blocked by this?

Reply:
This runs on the redis thread side, it won't block DM